apps/evm/lib/evm/message_call.ex

```diff
@@ -83,25 +83,47 @@ defmodule EVM.MessageCall do
     %{message_call | current_exec_env: exec_env}
   end
 
-  def update_state({n_gas, n_sub_state, n_exec_env, n_output}, message_call) do
-    if n_output != :failed do
-      machine_state = message_call.current_machine_state
-      exec_env = message_call.current_exec_env
-      {out_offset, _out_size} = message_call.output_params
-
-      updated_stack = Stack.push(machine_state.stack, 1)
-      machine_state = %{machine_state | stack: updated_stack, gas: machine_state.gas + n_gas}
-      machine_state = Memory.write(machine_state, out_offset, n_output)
-      exec_env = %{exec_env | account_interface: n_exec_env.account_interface}
-
-      %{
-        machine_state: machine_state,
-        exec_env: exec_env,
-        # https://github.com/poanetwork/mana/issues/153
-        sub_state: n_sub_state
-      }
-    else
-      failed_call(message_call)
+  def update_state({gas_remaining, n_sub_state, n_exec_env, output}, message_call) do
+    case output do
+      :failed ->
+        failed_call(message_call)
+
+      :invalid_input ->
```

Comment:
logic for `:invalid_input` and `_`  are almost exactly the same. Maybe we should move it to separate method with memory writing condition parameter?

Reply:
Yup, good call: [(d58c5) Combine two match branches for simplicity](https://github.com/poanetwork/mana/pull/220/commits/d58c50a6e91a7c3e42250265653c76cdbb226955)